### PR TITLE
Validate config.json

### PR DESCRIPTION
**Pull request by tomasz-stefaniak**
## Description

Validation of config.json

## Checklist

- [x] The base branch of this PR is `dev`, rather than `main`
- [ ] The relevant docs, if any, have been updated or created

## Screenshots

Screenshot 2024-10-15 at 10 15 53 PM

image

## Testing

[ For new or modified features, provide step-by-step testing instructions to validate the intended behavior of the change. ]

### CODE REVIEW

File: core/config/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * Validates a SerializedContinueConfig object to ensure all properties are correctly formed.
+ * @param config The configuration object to validate.
+ * @returns An array of error messages. If the array is empty, the config is valid.
+ */
+export function validateConfig(config: SerializedContinueConfig) {
+  const errors: ValidationErrorMessage[] = [];
+
+  // Validate models
+  if (!Array.isArray(config.models) || config.models.length === 0) {
```
Comment (sestinj) [inline core/config/validation.ts:17]: Don't worry about changing it in this PR, but just making the note here that we can switch to using zod later (definition is in packages/config-types/src/index.ts)

File: gui/src/components/Footer.tsx [hunk 2]
```diff
@@ -1,32 +1,38 @@
...
+    (store: RootState) => store.state.configError,
+  );
 
   function onClickMore() {
-    if (location.pathname === "/more") {
-      navigate("/");
-    } else {
-      navigate("/more");
-    }
+    navigate(pathname === ROUTES.MORE ? "/" : ROUTES.MORE);
```
Comment (sestinj) [inline gui/src/components/Footer.tsx:31]: Nice

**Review by sestinj — COMMENTED** [review]
I noticed a couple of things trying this out:
1) the warning icon turns out not to be super obvious. I think it is done perfectly for cases of non-fatal warnings, but if we have any errors that entirely break the config, then we might want to show it in a dialog or something. None of the current things seem necessarily critical though, so this can come in later work
2) it seems like the validation errors completely stop the config from loading, and I'm not actually sure that we want this. It's still possible to use the model, but it won't reflect the changes that the user thought they made in config.json. Especially when we want to return things that are just optional suggestions, we shouldn't stop the config from loading. This might mean we have to explicitly return rather than throwing and catching

File: core/config/validation.ts (hunk 1 above)
Comment (tomasz-stefaniak) [reply core/config/validation.ts:17]: I tested it - it's very easy to switch to zod, it took about 5 minutes of work, but it's a bit too eager and fails even valid configs. We'd need to look at the zod schema and see if it's correct.

**Comment by sestinj** [comment]
Sorry to be back and forth again on this—part of it is time zones—but I think we still need to both:
1) Give an extremely obvious/intrusive error dialog when a validation error is stopping the config from loading
and
2) Not stop the config from trying to load when we find validation errors

At least the latter is necessary when the validation errors are only "suggestion" status and not "critical". We should probably expand the `ValidationErrorMessage` type now to be an object with status and message to make sure we're doing this right

**Review by Thompson1985 — APPROVED** [review]

**Pull request merged**